### DISCUSSION TITLE
WebgalParser.parse 是否会立即执行各行对应的 gameScript

### DISCUSSION THREAD

**Discussion by linonetwo** [OpenWebGAL/WebGAL#499] [Q&A] [answered]
我看到 intro 函数里有 `WebGAL.events.userInteractNext.on(toNextIntroElement);` 开始监听事件，要执行到这一行需要先执行 intro 函数。

而在 startGame -> nextSentence 里，直接调用了 `WebGAL.events.userInteractNext.emit()` 发送事件，这说明 intro 应该是在此之前就执行了，所以我猜应该是在 startGame -> sceneParser 的 `WebgalParser.parse` 就在解析的时候立即执行了对应的方法吧。但实际看 `scriptParser` 里面并没有执行 `SCRIPT_CONFIG_MAP` 上的函数。

还有个疑问，这样如果有多行 intro，是不是 intro 会被执行多次，这样每个都会监听到事件，导致混乱？

如果 https://docs.openwebgal.com/tech/#%E5%87%86%E5%A4%87%E9%98%B6%E6%AE%B5-%E6%AD%A5%E8%BF%9B%E5%89%8D%E6%93%8D%E4%BD%9C 能指出是在哪个函数里有这些逻辑就好了

**Comment by MakinoharaShoko** [comment, marked as answer]
“而在 startGame -> nextSentence 里，直接调用了 WebGAL.events.userInteractNext.emit() 发送事件，这说明 intro 应该是在此之前就执行了”

错误的，intro 只有在游戏里有 intro 脚本且运行到该脚本才执行。这里的事件监听只是为了让 intro 可以在用户点击鼠标的时候提前显示下一行而已。

你会这样猜测是因为你认为事件监听一定在事件发出之前，所以 intro 执行一定在 nextSentence 执行之前。但实际上 nextSentence 函数，也是用户点击鼠标的回调。这个函数被称为“步进”，在用户点击鼠标，点击开始游戏，或者某些脚本执行完后，都有可能执行，并非只会执行一次。所以，在满足用户点击鼠标后可以让 intro 显示下一行这个功能的场景下，该事件回调只需在 intro 指令执行的时候注册即可，没有任何问题。

**Reply by linonetwo** [reply]
我看懂了，第一次由 startGame -> nextSentence 调用的 `WebGAL.events.userInteractNext.emit()` 可能不会有效果，因为此时还没调过 intro，不过接下来的 nextSentence -> runScript 就要调用 intro 了，第一次注册回调。

此后的用户点击 -> nextSentence 就会触发前一次 intro 注册的回调了。

![图片](https://github.com/OpenWebGAL/WebGAL/assets/3746270/0044ee17-b95a-4ef7-8a62-cc3e3e1ef943)
